// conventional-changelog-beemo/lib/index.js
"use strict";
/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const checkCommitFormat_1 = __importDefault(require("./checkCommitFormat"));
exports.checkCommitFormat = checkCommitFormat_1.default;
const getTypeGroup_1 = __importDefault(require("./getTypeGroup"));
exports.getTypeGroup = getTypeGroup_1.default;
const conventionalChangelog_1 = __importDefault(require("./conventionalChangelog"));
const parserOpts_1 = __importDefault(require("./parserOpts"));
const recommendedBumpOpts_1 = __importDefault(require("./recommendedBumpOpts"));
const writerOpts_1 = __importDefault(require("./writerOpts"));
__export(require("./constants"));
exports.config = Promise.resolve({
    conventionalChangelog: conventionalChangelog_1.default,
    parserOpts: parserOpts_1.default,
    recommendedBumpOpts: recommendedBumpOpts_1.default,
    writerOpts: writerOpts_1.default,
});
