// happo.io/build/DEFAULTS.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.customizeWebpackConfig = customizeWebpackConfig;
exports.pages = exports.compareThreshold = exports.githubApiUrl = exports.renderWrapperModule = exports.tmpdir = exports.publicFolders = exports.prerender = exports.plugins = exports.type = exports.configFile = exports.asyncTimeout = exports.targets = exports.project = exports.setupScript = exports.stylesheets = exports.include = exports.endpoint = exports.apiSecret = exports.apiKey = void 0;

var _os = _interopRequireDefault(require("os"));

var _path = _interopRequireDefault(require("path"));

var _RemoteBrowserTarget = _interopRequireDefault(require("./RemoteBrowserTarget"));

var _createHash = _interopRequireDefault(require("./createHash"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const cwdUniqueFolderName = `happo-tmp-${(0, _createHash.default)(process.cwd()).slice(0, 5)}`;
const apiKey = process.env.HAPPO_API_KEY;
exports.apiKey = apiKey;
const apiSecret = process.env.HAPPO_API_SECRET;
exports.apiSecret = apiSecret;
const endpoint = 'https://happo.io';
exports.endpoint = endpoint;
const include = '**/@(*-happo|happo).@(js|jsx|ts|tsx)';
exports.include = include;
const stylesheets = [];
exports.stylesheets = stylesheets;
const setupScript = undefined;
exports.setupScript = setupScript;
const project = undefined;
exports.project = project;
const targets = {
  chrome: new _RemoteBrowserTarget.default('chrome', {
    viewport: '1024x768'
  })
};
exports.targets = targets;
const asyncTimeout = 200;
exports.asyncTimeout = asyncTimeout;
const configFile = './.happo.js';
exports.configFile = configFile;
const type = 'react';
exports.type = type;
const plugins = [];
exports.plugins = plugins;
const prerender = true;
exports.prerender = prerender;
const publicFolders = [];
exports.publicFolders = publicFolders;

const tmpdir = _path.default.join(_os.default.tmpdir(), cwdUniqueFolderName);

exports.tmpdir = tmpdir;
const renderWrapperModule = undefined;
exports.renderWrapperModule = renderWrapperModule;
const githubApiUrl = 'https://api.github.com';
exports.githubApiUrl = githubApiUrl;
const compareThreshold = undefined;
exports.compareThreshold = compareThreshold;
const pages = undefined;
exports.pages = pages;

function customizeWebpackConfig(config) {
  // provide a default no-op for this config option so that we can assume it's
  // always there.
  return config;
}
