// happo.io/build/makeRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = makeRequest;

var _asyncRetry = _interopRequireDefault(require("async-retry"));

var _jsonwebtoken = _interopRequireDefault(require("jsonwebtoken"));

var _requestPromiseNative = _interopRequireDefault(require("request-promise-native"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function makeRequest(requestAttributes, {
  apiKey,
  apiSecret,
  maxTries = 0
}) {
  return (0, _asyncRetry.default)(async () => {
    const signed = _jsonwebtoken.default.sign({
      key: apiKey
    }, apiSecret, {
      header: {
        kid: apiKey
      }
    });

    const result = await (0, _requestPromiseNative.default)(Object.assign({
      gzip: true,
      auth: {
        bearer: signed
      }
    }, requestAttributes));
    return result;
  }, {
    retries: maxTries,
    onRetry: () => {
      const {
        method,
        url
      } = requestAttributes;
      console.warn(`Failed ${method} ${url}. Retrying...`);
    }
  });
}
