// @beemo/driver-eslint/lib/ESLintDriver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const event_1 = require("@boost/event");
const core_1 = require("@beemo/core");
// Success: Writes warnings to stdout
// Failure: Writes to stdout and stderr
class ESLintDriver extends core_1.Driver {
    constructor() {
        super(...arguments);
        this.onCreateIgnoreFile = new event_1.Event('create-ignore-file');
        /**
         * If an "ignore" property exists in the ESLint config, create an ".eslintignore" file.
         */
        this.handleCreateIgnoreFile = (context, configPath, config) => {
            if (!config.ignore) {
                return;
            }
            if (!Array.isArray(config.ignore)) {
                throw new TypeError(this.tool.msg('errors:eslintIgnoreInvalid'));
            }
            const ignorePath = configPath.parent().append('.eslintignore');
            const { ignore = [] } = config;
            this.onCreateIgnoreFile.emit([context, ignorePath, { ignore }]);
            fs_1.default.writeFileSync(ignorePath.path(), ignore.join('\n'));
            // Add to context so that it can be automatically cleaned up
            context.addConfigPath('eslint', ignorePath);
            // Delete the property else ESLint throws an error
            delete config.ignore;
        };
    }
    bootstrap() {
        this.setMetadata({
            bin: 'eslint',
            configName: '.eslintrc.js',
            description: this.tool.msg('app:eslintDescription'),
            title: 'ESLint',
        });
        this.onCreateConfigFile.listen(this.handleCreateIgnoreFile);
    }
    /**
     * ESLint writes warnings to stdout, so we need to display
     * both stdout and stderr on failure.
     */
    processFailure(error) {
        const { stderr, stdout } = error;
        if (stderr) {
            this.tool.console.logError(stderr);
        }
        if (stdout) {
            this.tool.console.log(stdout);
        }
    }
}
exports.default = ESLintDriver;
